### ISSUE TITLE
NNI 2021 June~July Iteration Planning

### ISSUE THREAD

**Issue by kvartet** [microsoft/nni#3724]
This is the plan for the iteration in June~July Iteration, it's a 6 weeks iteration.

# Release Plan for v2.4
- Release manager: @QuanluZhang 
- Feature freeze date: July 14
- Code freeze and demo date: July ~21~ ~23~ 26
- Branch cut and next release planning date: ~July 28~ August 2
- Release date: August ~2~ ~4~ 11

### Retiarii & NAS

- [x] P0 - NAS visualization
  - [x] trial detail/content re-format @ultmaster 6.29 (#3878)
  - [x] integrate Netron @98may @ultmaster 
- [ ] P0 - Add examples for NAS-Bench-series @ultmaster 
  - [x] 101 6.25 #3871
  - [x] 201 #3920 
  - [ ] P2 - 301
- [ ] P2 - Evaluate existing strategies (multi-trial) on NAS benchmarks @ultmaster 
- [ ] P0 - Support one system-level optimization: CSE optimization to deduplicate computation @hzhua pr #3202 
- [ ] P0 - *Improve training service to support CSE optimization* @hzhua #3897 
- [ ] P2 - TensorFlow support: graph converter and support inline mutation API @liuzhe-lz 
  - [ ] add raw framework
  - [ ] add example 7.12
- [ ] P0 - Migrate FBNet from NAS v1.0 to Retiarii framework @Ali 
- [ ] P2 - AutoDropout @skeletondyh 7.16
- [x] P0 - Searching for activation functions. @QuanluZhang ~6.24~ 6.28 (#3868 )
  - [x] P0 - (discussion) support autoactivation on base execution engine 6.29
- [x] P1 - Support PyTorch 1.8/1.9 @QuanluZhang 7.14 #3937 
- [x] P0 - Deal with retiarii experiment exit failure @QuanluZhang ~7.1~ 7.2 #3899 
- [x] P1 - survey serialize approaches @ultmaster ~7.15~ 7.19
- [x] P0 - support nn-meter in multi-trial nas @ultmaster @jianyu #3938 

### Model Compression

- [x] P0 -  Model Speedup Refactor @zheng-ningxin  pr #3462 
- [x] P1 - Add more examples for model speedup @zheng-ningxin #3880 
- [x] P0 - Add test for TensorRT @linbinskn 6.25 @J-shang 
- [x] P0 - Global sort for taylor @linbinskn 7.2 #3896 
  - [ ] support constraint
- [ ] P0 - support iterative pruning @J-shang  6.18 review meeting (done) / 6.23 sample / 6.25 7.14
  - [ ] P0 - implement model compression 2.0, include 8 maskers (basic pruners) 7.9 #3917 
  - [ ] P0 - improve pruners to more faithfully reproduce the papers @J-shang 
  - [ ] P2 - Refactor `compress()` in IterativePruner, unify support AGP, ADMM. (#3669 )
  - [ ] P2 - Add more criteria(strategy, i.e., acc drop toleronce) (discussion needed)
- [x] P0 - fix exclude not support in partial compressor configlist @J-shang pr ready #3815 
- [x] P0 - review meeting of model compression @QuanluZhang 6.22
- [x] P1 - Add one compression algorithm for Bert/Transformer @xiaowu0162 
  - [x] review meeting 6.23 
  - [x] first version implementation ~7.1~ 7.9 #3884 

### Training service & NNI manager && nnictl
- [x] P0 - file name validation, from error to warning and improve warning message @liuzhe-lz (#3843)
- [x] P0 - JuypterLab support - #3668 @liuzhe-lz 
  - [x] - basic iframe, integrate installation script ~6.29~ 7.14 #3954 
- [x] P0 - Add test for shared storage (PR #3455) @acured / @QuanluZhang (NFS storage)
  - [x] pipeline remote linux-linux timeout @ultmaster @SparkSnail 
  - [x] add NFS test ~7.6~ 7.7 #3913 
- [ ] P0 - (discussion needed) design an API to share data loading/preprocessing among trials (#3294 )
  - [x] design discussion
  - [x] design review @acured 7.2
  - [x] follow-up discussion ~7.9~ 7.12
  - [ ] P2 - implementation
- [x] P0 - View experiment from experiment folder @SparkSnail 6.25 #3870 
- [ ] P0 - Make **reuse** default @SparkSnail 
     - [x] P0 - Hybrid kubeflow support (migrate kubeflow to environment service) @SparkSnail ~7.2~ 7.5 #3919 
     - [ ] P1 - support storage service for different training services, review and implementation
     - [ ] P1 - add more test for remote mode training service
     - [ ] P2 - Frameworkcontroller support @SparkSnail
- [x] P0 - update PT support @SparkSnail ~7.9~ 7.12
- [x] P0 - update quickstart doc to config v2 @kvartet #3861 
- [x] P0 - update doc for launch from python: API reference @kvartet 6.29 #3885 @ultmaster 
- [x] P0 - support view trial log on webui for experiment view @SparkSnail #3872 
- [ ] P0 - fix path escape error in windows (issue #3869) @liuzhe-lz @acured @J-shang #3898
  - [x] P(-1) - fix nnimanager ut framework @liuzhe-lz 7.12 #3933 
  - [x] P(-1) - fix nnimanager ut test cases: ?, ?, ? @liuzhe-lz #3933 
  - [ ] add back removed ut
- [x] P1 - (design) backend design for project/management @acured 7.5
- [x] P0 - update doc for exp config v2 @liuzhe-lz ~7.12~ ~7.14~ 7.19 #3966 
- [x] P0 - fix training service reuse mode bug @hzhua #3941 
- [x] P0 - fix IP address obtaining bug: EBADF, in reuse mode? @liuzhe-lz 7.19

### WebUI
- [x] P0 - node upgrade @liuzhe-lz (#3828 pr ready)
  - [x] Change IP detection to modern way (#3860)
- [x] P0 - demo: Experiment/Project Management @Lijiaoa 6.25
- [x] P0 - Show dict intermediate result in the dialog of single trial intermediate result @Lijiaoa 7.2 #3895 
- [ ] P0 - Adaptive UI (height) for high resolution, current window for high resolution need improvement @junzhuzhang  @yangou1988 @Lijiaoa 
  - [x] fix bug in table search 7.14 #3935 
  - [ ] overview page: blank at the bottom
  - [ ] align issue in overview page
  - [ ] support searching customized string in the search box

### Hyper-parameter tuning

- [x] P0 - Test FLAML with HPO benchmark @xiaowu0162 6.25
- [x] P1 - Improve HPO benchmarks, add more concrete conclusions @xiaowu0162 7.12 #3925 
  - [ ] add one benchmark and one search space 7.14 #3925 
- [ ] P1 - support nested search space @kvartet 6.28 discussion meeting
- [x] P1 - support time limit early stop algorithm (#3740) @acured 6.25 #3863 
- [ ] P1 - Improve hyperband, PBT with shared storage (dependency on shared storage implementation) @acured 
- [ ] P1 - upgrade cuda, config_space, docker image  @liuzhe-lz 7.21 #3976
- [x] P1 - search space type validation, check hyper-parameter's type, if not support expose warning @liuzhe-lz 7.21 #3975 
- [ ] P1 - grid search does not support `_name` for nested search space (update doc) @kvartet 
- [ ] P1 - evolution tuner does not mutate sub search space for nested search space @kvartet 

### Pipeline

- [x] P0 - support python 3.9 in pipeline (wait for pytorch upgrade from 1.6 to 1.9) @ultmaster #3881 
  - [x] fix local pipeline nnimanager ut (nnimanager test can be added back, ipc test cannot be added back) @J-shang #3970 
  - [ ] P2 - write ipc test
- [x] P0 - code coverage @ultmaster 6.25 
- [x] P0 - fix pipeline error: disk full, docker images are in primary disk
  - [x] update crontab command @J-shang
  - [x] config docker image place @SparkSnail 
- [x] P0 - fix windows pipeline @ultmaster #3926 
  - [x] fix linux
- [ ] P0 - fix and improve pipeline
  - [x] fast test: lint, UT, simple IT, timeout: (1) blocked by certain test (2) real timeout -> upgrade machine @ultmaster  #3962 
  - [ ] nnimanager ut, nni-ci-gpu-03 failed, maybe due to this machine's environment, cannot be executed through daemon, need reimage -> create a new vm
  - [x] review speedup test, why it requires large memory @J-shang 

### NAS

- [ ] P1 - Evaluate one-shot strategies on NAS benchmarks @ultmaster 
- [ ] P1 - Further improve the logic of graph generation and code generation (Further integrated with TorchScript) @QuanluZhang
- [ ] P1 - Test classic nas tuners on NNI NAS benchmark and report evaluation results @ultmaster 
- [ ] P2 - Migrate Cream from NAS v1.0 to Retiarii framework @ yuhangchen
- [ ] P2 - refactor of Retiarii execution engine @ultmaster 
  - [ ] support local debug mode for pure-python execution engine
  - [ ] support weight transfer from supernet to submodel
  - [ ] support export of top model for pure-python execution engine
  - [ ] P2 - refactor of "self._cur_samples" in mutators
  - [ ] P2 - support validation for other strategies
- [ ] P1 - Review strategy/experiment/rest stop condition @QuanluZhang 
- [ ] P1 - support graph-based NAS algorithms @ultmaster @skeletondyh
- [ ] P1 - Deal with strategy failure 

### model compression
- [ ] P1 - MixedMaskerPruner @J-shang 
- [ ] P2 - Refactor of model graph generation (as an independent component) @zheng-ningxin 

### nnimanager
- [ ] P2 - Mask confidential config fields (depends on prev items) @liuzhe-lz
- [ ] P2 - `nnictl experiment export` improvements @J-shang 
- [ ] P2 - internal tuner APIs for improving tuner debuggability, allow debugger tool @liuzhe-lz 
- [ ] P2 - nnictl auto-completion @liuzhe-lz 
- [ ] P2 - Dependency-aware supported in iterative pruner (discussion needed)
- [ ] P2 - Add full test (l2filter, Taylor, NetAdapt, SimulatedAnnealing, AutoCompress, AMC, Sensitivity, ADMM, Lottery Ticket)
- [ ] P1 - reuse mode: ADL kubeflow support @SparkSnail
- [ ] P2 - add/del tag of experiments (pending support from rest server)
- [ ] UI: support experiment resume/view/stop on webUI
- [ ] P0 - (discussion needed) how to support specifying initial configs (#3801 ) @ultmaster 
- [ ] P2 - (discussion needed) support constraints on hyper-parameters @ultmaster 
- [ ] P2 - handle all reject errors @J-shang

**Comment by QuanluZhang** [comment]
From Ali PAI team
- [ ] global sort, importance sort on all layers
- [ ] QAT improvement, PTQ initialization (set scale, zero point), post-training quantization
- [ ] new model compression algorithms (e.g., Bert), model-specific compression (not that generic), sparsifying training and sparsifying operator

**Comment by QuanluZhang** [comment]
bugbash

### Retiarii & NAS

- [ ] P0 - NAS visualization @ultmaster 6.29 (#3878) tester: @jonsnows
- [ ] P0 - Add examples for NAS-Bench-series @ultmaster @skeletondyh 
  - [ ] 101 6.25 #3871
  - [ ] 201 #3920 
- [ ] P0 - Support one system-level optimization: CSE optimization to deduplicate computation @hzhua pr #3202 tester: @QuanluZhang 
- [ ] P0 - Improve training service to support CSE optimization @hzhua #3897 tester: @QuanluZhang 
- [ ] P0 - Searching for activation functions. @QuanluZhang ~6.24~ 6.28 (#3868 ) tester: @ultmaster 
  - [ ] P0 - (discussion) support autoactivation on base execution engine 6.29
- [ ] P1 - Support PyTorch 1.8/1.9 @QuanluZhang 7.14 #3937 tester: @ultmaster 
- [ ] P0 - Deal with retiarii experiment exit failure @QuanluZhang ~7.1~ 7.2 #3899 tester: @ultmaster 
- [ ] P0 - support nn-meter in multi-trial nas @ultmaster @jianyu #3938 tester: @zheng-ningxin 

### Model Compression

- [x] P0 -  Model Speedup Refactor @zheng-ningxin  pr #3462 tester: @J-shang 
- [x] P1 - Add more examples for model speedup @zheng-ningxin #3880 tester: @linbinskn 
- [x] P0 - Add test for TensorRT @linbinskn 6.25 @J-shang tester: @xiaowu0162 
- [x] P0 - Global sort for taylor @linbinskn 7.2 #3896 tester: @xiaowu0162 
  - [ ] support constraint
- [ ] P0 - implement model compression 2.0, include 8 maskers (basic pruners) 7.9 #3917 tester: @linbinskn 
- [x] P0 - fix exclude not support in partial compressor configlist @J-shang pr ready #3815 tester: @xiaowu0162 
- [ ] P1 - Add one compression algorithm for Bert/Transformer @xiaowu0162 tester: @J-shang 
  - [ ] first version implementation ~7.1~ 7.9 #3884 

### Training service & NNI manager && nnictl
- [x] P0 - file name validation, from error to warning and improve warning message @liuzhe-lz (#3843) tester: @acured 
- [x] P0 - JuypterLab support - #3668 @liuzhe-lz tester: @SparkSnail 
  - [x] - basic iframe, integrate installation script ~6.29~ 7.14 #3954 
- [x] P0 - test shared storage (PR #3455) @acured / @QuanluZhang (NFS storage) 
  - [ ] local, remote,  tester: @QuanluZhang 
  - [ ] aml, pai,  tester: @SparkSnail 
- [x] P0 - View experiment from experiment folder @SparkSnail 6.25 #3870 @liuzhe-lz 
- [ ] P0 - Make **reuse** default @SparkSnail 
     - [ ] P0 - Hybrid kubeflow support (migrate kubeflow to environment service) @SparkSnail ~7.2~ 7.5 #3919 tester: @acured 
- [x] P0 - update PT support @SparkSnail ~7.9~ 7.12 tester: @SparkSnail 
- [x] P0 - update quickstart doc to config v2 @kvartet #3861 tester: @QuanluZhang 
- [x] P0 - update doc for launch from python: API reference @kvartet 6.29 #3885 @ultmaster tester: @QuanluZhang 
- [x] P0 - support view trial log on webui for experiment view @SparkSnail #3872 tester: @liuzhe-lz 
- [ ] P0 - fix path escape error in windows (issue #3869) @liuzhe-lz @acured @J-shang #3898 tester: @ultmaster 
- [x] P0 - fix training service reuse mode bug @hzhua #3941 tester: @acured 
- [x] P0 - fix IP address obtaining bug: EBADF, in reuse mode? @liuzhe-lz 7.19 tester: @acured 
  - [x] Change IP detection to modern way (#3860)

### WebUI
- [x] P0 - Show dict intermediate result in the dialog of single trial intermediate result @Lijiaoa 7.2 #3895 tester: @liuzhe-lz 
- [x] fix bug in table search 7.14 #3935 tester: @liuzhe-lz 

### Hyper-parameter tuning

- [ ] P1 - Improve HPO benchmarks, add more concrete conclusions @xiaowu0162 7.12 #3925 
  - [ ] add one benchmark and one search space 7.14 #3925 tester: @jonsnows 
- [x] P1 - support time limit early stop algorithm (#3740) @acured 6.25 #3863 tester: @skeletondyh 
- [ ] P1 - search space type validation, check hyper-parameter's type, if not support expose warning @liuzhe-lz 7.21 #3975 tester: @xiaowu0162

**Comment by QuanluZhang** [comment]
fix pipeline:
- [ ] integration test windows to linux @ultmaster 
- [ ] integration test pai linux @liuzhe-lz 
- [ ] gpu-04 reconfig disks @liuzhe-lz 
- [ ] integration test hybrid @SparkSnail 
- [ ] full test linux @ultmaster
